src/cursor/common.rs

```diff
@@ -98,63 +105,131 @@ where
+                        // connection to the pool.
+                        self.pinned_connection = PinnedConnection::Unpinned;
+                    }
+                    if let Err(e) = &result {
+                        if e.is_network_error() {
+                            // Flag the connection as invalid, preventing a killCursors command,
+                            // but leave the connection pinned.
+                            self.pinned_connection.invalidate();
                         }
-
-                        self.exhausted = exhausted;
-                        self.provider.clear_execution(session, exhausted);
-                        self.buffer = result?.batch;
                     }
-                    Poll::Pending => return Poll::Pending,
+
+                    self.exhausted = exhausted;
+                    self.provider.clear_execution(session, exhausted);
+                    let result = result?;
+                    self.buffer = result.batch;
+                    self.post_batch_resume_token = result.post_batch_resume_token;
                 }
+                Poll::Pending => return Poll::Pending,
             }
+        }
 
-            match self.buffer.pop_front() {
-                Some(doc) => {
-                    return Poll::Ready(Some(Ok(bson::from_slice(doc.as_bytes())?)));
-                }
-                None if !self.exhausted && !self.pinned_connection.is_invalid() => {
-                    self.start_get_more();
-                }
-                None => return Poll::Ready(None),
+        match self.buffer.pop_front() {
+            Some(doc) => {
+                return Poll::Ready(Ok(BatchValue::Some {
+                    doc,
+                    is_last: self.buffer.is_empty(),
+                }));
+            }
+            None if !self.exhausted && !self.pinned_connection.is_invalid() => {
+                self.start_get_more();
+                return Poll::Ready(Ok(BatchValue::Empty));
             }
+            None => return Poll::Ready(Ok(BatchValue::Exhausted)),
         }
     }
 }
 
+// To avoid a private trait (`CursorStream`) in a public interface (`impl Stream`), this is provided
+// as a free function rather than a blanket impl.
+pub(crate) fn stream_poll_next<S, V>(this: &mut S, cx: &mut Context<'_>) -> Poll<Option<Result<V>>>
+where
+    S: CursorStream,
+    V: for<'a> serde::Deserialize<'a>,
+{
+    loop {
+        match this.poll_next_in_batch(cx) {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(bv) => match bv? {
+                BatchValue::Some { doc, .. } => {
+                    return Poll::Ready(Some(Ok(bson::from_slice(doc.as_bytes())?)))
+                }
+                BatchValue::Empty => continue,
+                BatchValue::Exhausted => return Poll::Ready(None),
+            },
+        }
+    }
+}
+
+pub(crate) struct NextInBatchFuture<'a, T>(&'a mut T);
```

Comment:
It seems awkward to have to have a one-off struct to go from the desugared `fn() -> Poll<T>` form back to `.await`, but I guess this is the way until `poll_fn` stabilizes.